// experimental: Port single click Event OnClick(shift,atrl,alt) from scite.ru to scintilla Mainline.
diff -Nurw scintilla/include/Scintilla.h scintilla_mod/include/Scintilla.h
--- scintilla/include/Scintilla.h	2017-05-26 00:25:00.000000000 +0200
+++ scintilla_mod/include/Scintilla.h	2017-11-15 18:28:50.000000000 +0100
@@ -1010,6 +1010,7 @@
 #define SCI_SETIDENTIFIERS 4024
 #define SCI_DISTANCETOSECONDARYSTYLES 4025
 #define SCI_GETSUBSTYLEBASES 4026
+#define SCI_SETMOUSECAPTURE 4061
 #define SC_MOD_INSERTTEXT 0x1
 #define SC_MOD_DELETETEXT 0x2
 #define SC_MOD_CHANGESTYLE 0x4
@@ -1102,6 +1103,8 @@
 #define SCN_FOCUSOUT 2029
 #define SCN_AUTOCCOMPLETED 2030
 #define SCN_MARGINRIGHTCLICK 2031
+#define SCN_CLICK 2061
+
 /* --Autogenerated -- end of section automatically generated from Scintilla.iface */
 
 /* These structures are defined to be exactly the same shape as the Win32
diff -Nurw scintilla/include/Scintilla.iface scintilla_mod/include/Scintilla.iface
--- scintilla/include/Scintilla.iface	2017-05-26 00:25:00.000000000 +0200
+++ scintilla_mod/include/Scintilla.iface	2017-11-15 19:33:56.000000000 +0100
@@ -4859,7 +4865,8 @@
 # There are no provisional APIs currently, but some arguments to SCI_SETTECHNOLOGY are provisional.
 
 cat Provisional
-
+evt void Click=2061(int modifiers)
+set void SetMouseCapture=4061(bool on,)
 cat Deprecated
 
 # Deprecated in 3.5.5
diff -Nurw scintilla/src/Editor.cxx scintilla_mod/src/Editor.cxx
--- scintilla/src/Editor.cxx	2017-05-26 00:25:02.000000000 +0200
+++ scintilla_mod/src/Editor.cxx	2017-11-15 19:29:02.000000000 +0100
@@ -2331,6 +2331,16 @@
 	NotifyParent(scn);
 }
 
+void Editor::NotifyClick(Point pt, bool shift, bool ctrl, bool alt) {
+	SCNotification scn = {};
+	scn.nmhdr.code = SCN_CLICK;
+	scn.line = LineFromLocation(pt);
+	scn.position = PositionFromLocation(pt, true);
+	scn.modifiers = (shift ? SCI_SHIFT : 0) | (ctrl ? SCI_CTRL : 0) |
+		(alt ? SCI_ALT : 0);
+	NotifyParent(scn);
+}
+
 void Editor::NotifyDoubleClick(Point pt, int modifiers) {
 	SCNotification scn = {};
 	scn.nmhdr.code = SCN_DOUBLECLICK;
@@ -4433,7 +4443,7 @@
 
 void Editor::ButtonDownWithModifiers(Point pt, unsigned int curTime, int modifiers) {
 	SetHoverIndicatorPoint(pt);
-	//Platform::DebugPrintf("ButtonDown %d %d = %d alt=%d %d\n", curTime, lastClickTime, curTime - lastClickTime, alt, inDragDrop);
+	//Platform::DebugPrintf("ButtonDown %d %d = %d %d\n", curTime, lastClickTime, curTime - lastClickTime, inDragDrop);
 	ptMouseLast = pt;
 	const bool ctrl = (modifiers & SCI_CTRL) != 0;
 	const bool shift = (modifiers & SCI_SHIFT) != 0;
@@ -4444,6 +4454,7 @@
 	newCharPos = MovePositionOutsideChar(newCharPos, -1);
 	inDragDrop = ddNone;
 	sel.SetMoveExtends(false);
+	bool notifyClick = false;
 
 	if (NotifyMarginClick(pt, modifiers))
 		return;
@@ -4462,7 +4473,7 @@
 		SetSelection(newPos);
 	}
 	if (((curTime - lastClickTime) < Platform::DoubleClickTime()) && Close(pt, lastClick, doubleClickCloseThreshold)) {
-		//Platform::DebugPrintf("Double click %d %d = %d\n", curTime, lastClickTime, curTime - lastClickTime);
+		//Platform::DebugPrintf("Double click %d %d = %d\n", curTime, lastClickTime, curTime - lastClickTime);
 		SetMouseCapture(true);
 		if (FineTickerAvailable()) {
 			FineTickerStart(tickScroll, 100, 10);
@@ -4612,12 +4623,14 @@
 				sel.Rectangular() = SelectionRange(newPos, anchorCurrent);
 				SetRectangularRange();
 			}
+			notifyClick = true;
 		}
 	}
 	lastClickTime = curTime;
 	lastClick = pt;
 	lastXChosen = static_cast<int>(pt.x) + xOffset;
 	ShowCaretAtCurrentPosition();
+	if (notifyClick) NotifyClick(pt, shift, ctrl, alt);
 }
 
 void Editor::RightButtonDownWithModifiers(Point pt, unsigned int, int modifiers) {
@@ -8170,6 +8183,13 @@
 		pdoc->ChangeLexerState(static_cast<int>(wParam), static_cast<int>(lParam));
 		break;
 
+	case SCI_SETMOUSECAPTURE:
+		SetMouseCapture(wParam != 0);
+				vs.viewEOL = wParam != 0;
+		InvalidateStyleRedraw();
+
+		break;
+		
 	case SCI_SETIDENTIFIER:
 		SetCtrlID(static_cast<int>(wParam));
 		break;
diff -Nurw scintilla/src/Editor.h scintilla_mod/src/Editor.h
--- scintilla/src/Editor.h	2017-05-26 00:25:02.000000000 +0200
+++ scintilla_mod/src/Editor.h	2017-11-15 17:48:58.000000000 +0100
@@ -424,6 +424,7 @@
 	void NotifyChar(int ch);
 	void NotifySavePoint(bool isSavePoint);
 	void NotifyModifyAttempt();
+	virtual void NotifyClick(Point pt, bool shift, bool ctrl, bool alt);
 	virtual void NotifyDoubleClick(Point pt, int modifiers);
 	virtual void NotifyDoubleClick(Point pt, bool shift, bool ctrl, bool alt);
 	void NotifyHotSpotClicked(Sci::Position position, int modifiers);
